harvester/harvest.py

```diff
@@ -57,6 +58,126 @@
+            SynchronizeException: Any ckan sync issue
+        """
+        if record.status == "success" or record.status == "error":
+            logger.info(
+                f"{record.identifier} has status '{record.status}'. bypassing {record.action}"
+            )
+            return
+
+        # pre-sync try to ckanify record metadata
+        try:
+            if record.action == "create" or record.action == "update":
+                self.ckanify_record(record)
+        except Exception as e:
+            record.status = "error"
+            raise DCATUSToCKANException(
+                repr(e), record.harvest_source.job_id, record.id
+            )
+
+        # begin ckanpi sync proper
+        start = get_datetime()
+        try:
+            if record.action == "delete":
+                self.delete_record(record)
+            elif record.action == "create":
+                self.ckanify_record(record)
+                res = self.create_record(record)
+                record.ckan_id = res["id"]
+                record.ckan_name = record.ckanified_metadata["name"]
+            elif record.action == "update":
+                self.ckanify_record(record)
+                self.update_record(record)
+        except Exception as e:
+            record.status = "error"
+            raise SynchronizeException(
+                f"failed to {record.action} for {record.identifier} :: {repr(e)}",
+                record.harvest_source.job_id,
+                record.id,
+            )
+
+        logger.info(
+            f"time to {record.action} {record.identifier} \
+                {get_datetime() - start}"
+        )
+        record.status = "success"
+
+        # update harvest reporter
+        record.harvest_source.reporter.update(record.action)
+
+        # # update harvest job
+        record.harvest_source.db_interface.update_harvest_job(
+            record.harvest_source.job_id, record.harvest_source.reporter.report()
+        )
+
+        return True
+
+    def create_record(self, record, retry=False) -> dict:
+        from harvester.utils.ckan_utils import add_uuid_to_package_name
+
+        try:
+            return ckan.action.package_create(**record.ckanified_metadata)
+        except Exception as e:
+            if retry is False:
+                record.ckanified_metadata["name"] = add_uuid_to_package_name(
+                    record.ckanified_metadata["name"]
+                )
+                return self.create_record(record, retry=True)
+            else:
+                raise e
+                # will be caught by outer SynchronizeException
+
+    def update_record(self, record) -> dict:
+        updated_metadata = {
+            **record.ckanified_metadata,
+            **{"id": record.ckan_id, "name": record.ckan_name},
+        }
+        return ckan.action.package_update(**updated_metadata)
+
+    def delete_record(self, record) -> None:
+        return ckan.action.dataset_purge(**{"id": record.ckan_id})
```

Comment:
suggestion (non-blocking)**: I think that with just one keyword argument, it's just as easy to not use the dict unwrapping:
```suggestion
        return ckan.action.dataset_purge(id=record.ckan_id)
```